core/src/main/java/org/elasticsearch/cluster/metadata/PersistedIndexMetaData.java

```diff
@@ -0,0 +1,185 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.ParseFieldMatcher;
+import org.elasticsearch.common.xcontent.FromXContentBuilder;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Objects;
+
+/**
+ * Index metadata that is persisted on disk.
+ */
+final public class PersistedIndexMetaData implements FromXContentBuilder<PersistedIndexMetaData>, ToXContent {
+
+    public static final PersistedIndexMetaData PROTOTYPE = new PersistedIndexMetaData(null, null);
+    public static final String INDEX_STATE_FILE_PREFIX = "state-";
+    public static final String CLUSTER_UUID_NA_VALUE = "_na_";
+    private static final String CLUSTER_UUID_FIELD = "clusterUUID";
+    private static final String INDEX_METADATA_FIELD = "indexMetaData";
+
+    private final IndexMetaData indexMetaData;
+    private final String clusterUUID;
+
+    public PersistedIndexMetaData(final IndexMetaData indexMetaData, final String clusterUUID) {
+        this.indexMetaData = indexMetaData;
+        this.clusterUUID = clusterUUID;
+    }
+
+    /**
+     * The cluster state's index metadata
+     */
+    public IndexMetaData getIndexMetaData() {
+        return indexMetaData;
+    }
+
+    /**
+     * The cluster UUID of the cluster that last created or updated the index
+     */
+    public String getClusterUUID() {
+        return clusterUUID;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+        if (other == null || this.getClass() != other.getClass()) {
+            return false;
+        }
+        final PersistedIndexMetaData that = (PersistedIndexMetaData) other;
+        return Objects.equals(clusterUUID, that.clusterUUID) && Objects.equals(indexMetaData, that.indexMetaData);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(clusterUUID, indexMetaData);
+    }
+
+    @Override
+    public PersistedIndexMetaData fromXContent(XContentParser parser, ParseFieldMatcher parseFieldMatcher) throws IOException {
+        if (parser.currentToken() == null) { // fresh parser? move to the first token
```

Comment:
@bleskes Thanks for alerting me to the ObjectParser.  In looking at it though, it may be tricky because during the parsing of the x-content, I have to make a decision whether to parse the old format (just IndexMetaData) or the new format (IndexStateMetaData).